server/Dockerfile

```diff
@@ -0,0 +1,57 @@
+# create base image with all packages, proper entrypoint, and directories created
+FROM node:dubnium-alpine AS base
+
+# install any packages we need from apt here
+RUN apk add --update dumb-init
+
+# set entrypoint to `dumb-init` as it handles being pid 1 and forwarding signals
+# so that you dont need to bake that logic into your node app
+ENTRYPOINT ["dumb-init", "--"]
```

Comment:
Can this be done with Docker's `--init` flag? https://docs.docker.com/engine/reference/run/#specify-an-init-process

Reply:
I believe `--init` is backed by `tini` which is an alternative to `dumb-init` so my guess is Yes, but maybe there are some feature behind `dumb-init` ?

(`--init` require Docker 1.13+)

Reply:
I would prefer to use docker features rather than add dependencies to the image where it's possible.  I'm comfortable with requiring that version of docker.

Reply:
I've always used dumb-init always tini, because i prefer signal rewriting to subprocess reaping within my containers.

https://gist.github.com/StevenACoffman/41fee08e8782b411a4a26b9700ad7af5

If tini is preferred because it's not an additional dependency, that's fine, just wanted to explain the _why_ for dumb-init.

>Dumb-init or Tini
>PID 1 is special in unix, and so omitting an init system often leads to incorrect handling of processes and signals, and can result in problems such as containers which can't be gracefully stopped, or leaking containers which should have been destroyed.
>In Linux, processes in a PID namespace form a tree with each process having a parent process. Only one process at the root of the tree doesn't really have a parent. This is the "init" process, which has PID 1.
>Processes can start other processes using the fork and exec syscalls. When they do this, the new process' parent is the process that called the fork syscall. fork is used to start another copy of the running process and exec is used to start different process. Each process has an entry in the OS process table. This records info about the process' state and exit code. When a child process has finished running, its process table entry remains until the parent process has retrieved its exit code using the wait syscall. This is called "reaping" zombie processes.

Zombie processes are processes that have stopped running but their process table entry still exists because the parent process hasn't retrieved it via the wait syscall. Technically each process that terminates is a zombie for a very short period of time but they could live for longer.

Something like tini or dumb-init can be used if you have a process that spawns new processes but doesn't have good signal handlers implemented to catch child signals and stop your child if your process should be stopped etc.

Bash scripts for example do NOT handle and emit signals properly.

Dumb-init supports signal rewriting and Tini doesn't, but Tini supports subreapers and they don't.

Reply:
It doesn't seem like we use signal rewriting anywhere here.  As far as I can tell both tini and dumb-init should be forwarding signals to the node server, and both should exit when the node server exits.  In other words, I would expect tini and dumb-init to have the exact same behavior unless 1. dumb-init is configured to map some exit codes (tini cannot do this, but we also aren't using this feature) or 2. the node server creates zombies by mistake (only tini would clean them up).  Does that seem right?

Reply:
So, I think that we should include either tini or dumb-init. Kubernetes which is _the_ docker deployment framework of choice generally doesnt support init flags for docker containers that i can see.

Reply:
Cool, that makes sense.  I'm down to leave it as-is—thanks for bearing with me.

Reply:
Absolutely - I knew docker had added the init flag since i started using dumb-init, but didnt realize that kube didnt support it yet. Always fun learning on the fly.